File: sdks/csnet/BaSys40.Utils/ResultHandling/Result.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.Serialization;

namespace BaSys40.Utils.ResultHandling
{
    public class Result : IResult
    {
        public bool Success { get; private set; }

        public object Entity { get; private set; }

        public Type EntityType { get; private set; }

        private List<IMessage> messages;

        public List<IMessage> Messages
        {
            get
            {
                if (this.messages == null)
                    this.messages = new List<IMessage>();
                return this.messages;
            }
        }

        public Result(bool success) : this(success, null, null, null)
        { }

        public Result(bool success, List<IMessage> messages) : this(success, null, null, messages)
        { }

        public Result(bool success, object entity, Type entityType) : this(success, entity, entityType, null)
        { }

        public Result(Exception e) : 
            this(false, e, e.GetType(), new List<IMessage>()
            {
                new Message(MessageType.Error, e.Message),
                new Message(MessageType.Error, e.InnerException != null ? e.InnerException.Message : "InnerException = null")
            })
        { }

        public Result(bool success, object entity, Type entityType, List<IMessage> messages)
        {
            Success = success;

            if (messages != null)
                foreach (Message msg in messages)
                    Messages.Add(msg);

            if (entity != null && entityType != null)
            {
                Entity = entity;
                EntityType = entityType;
            }

        }

        public T GetEntity<T>()
        {
            if (Entity != null && 
                (Entity is T ||
                Entity.GetType().IsAssignableFrom(typeof(T)) ||
                Entity.GetType().GetInterfaces().Contains(typeof(T))))
                return (T)Entity;
            return default(T);
        }
    }

    public class Result<TEntity> : Result, IResult<TEntity>
    {
        [IgnoreDataMember]
        public new TEntity Entity { get; private set; }

        public Result(bool success) : this(success, default(TEntity), new List<IMessage>())
        { }
        public Result(bool success, TEntity entity) : this(success, entity, new List<IMessage>())
        { }
        public Result(bool success, IMessage message) : this(success, default(TEntity), new List<IMessage>() { message })
        { }
        public Result(bool success, List<IMessage> messages) : this(success, default(TEntity), messages)
        { }
        public Result(bool success, TEntity entity, IMessage message) : this(success, entity, new List<IMessage>() { message })
        { }
        public Result(Exception e) : base(e)
        { }
        public Result(bool success, TEntity entity, List<IMessage> messages) : base(success, entity, typeof(TEntity), messages)
        {
            Entity = entity;
        }
    }
}

File: sdks/java/basys.sdk/src/main/java/org/eclipse/basyx/vab/backend/server/utils/Result.java
package org.eclipse.basyx.vab.backend.server.utils;

import java.util.Arrays;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;

/**
 * Wrapper class that handles meta-data
 * 
 * @author pschorn
 *
 */
@SuppressWarnings("serial")
public class Result  extends HashMap<String, Object> implements IResult {
		
	public static final String SUCCESS = "success";
	public static final String ISEXCEPTION = "isException";
	public static final String MESSAGES = "messages";
	public static final String ENTITY = "entity";
	public static final String ENTITYTYPE = "entityType";

	public Result(boolean success) {
		this(success, null, null, null);
	}
	
	public Result(boolean success, IMessage message) {
		this(success, null, null, new LinkedList<IMessage>(Arrays.asList(message)));
	}
	
	public Result(boolean success, List<IMessage> messages) {
		this(success, null, null, messages);
	}
	
	public Result(boolean success, Object entity,  Class<?> entityType, List<IMessage> messages) {
		put(SUCCESS, success);
		
		if (messages != null) {
			
			List<Map<String, Object>> messageslist = new LinkedList<Map<String, Object>> ();
			for (IMessage msg : messages) {
				
				MessageType type = msg.getMessageType();
				
				if (type.equals(MessageType.Exception)) {
					put(ISEXCEPTION, true); // make sure isException is set!
				}
				
				messageslist.add((Message) msg);
			}
			
			put(MESSAGES, messageslist);
		}
		
		if (entity != null && entityType != null) {
			put(ENTITY, entity);
			put(ENTITYTYPE, entityType.getName());
		}
	}
	
	public Result(IResult result) {
		this(result.success(), result.getEntity(), result.getEntityType(), result.getMessages());
	}
	
	public Result(Exception e) {
		this(false, getMessageListFromException(e));
	}
	
	
	private static List<IMessage> getMessageListFromException(Exception e){
		
		List<IMessage> messageList = new LinkedList<IMessage>();
		
		
		if (e.getCause() != null) {
			messageList.addAll(getMessageListFromException((Exception) e.getCause())); 
		}
		
		// prepare stacktrace
		String trace = "\n";
		for (StackTraceElement s : e.getStackTrace()) {

			trace = trace + " at " + s.toString() + "\n";
		}

		messageList.add(
				new Message(MessageType.Exception, e.getClass().getName() + ": " + trace)); // replace
																															// with
																															// desired
																															// debugging
																															// output
		
		return messageList;
	}

	@Override
	public Class<?> getEntityType() {
		return (Class<?>) get(ENTITYTYPE);
	}

	@Override
	public Object getEntity() {
		return get(ENTITY);
	}

	@Override
	public boolean success() {
		return (boolean) get(SUCCESS);
	}

	@Override
	public boolean isException() {
		return (boolean) get(ISEXCEPTION);
	}

	@SuppressWarnings("unchecked")
	@Override
	public List<IMessage> getMessages() {
		return (List<IMessage>) get(MESSAGES);
	}

}
